geometric_kernels/spaces/so.py

```diff
@@ -0,0 +1,368 @@
+            def xi1(qs):
+                mat = sympy.Matrix(rank, rank, lambda i, j: gammas[i] ** qs[j] - gammas_conj[i] ** qs[j])
+                return sympy.Poly(sp_det(mat, method='berkowitz'), chi_variables)
+            qs = [pk + rank - k - 1 if k != rank - 1 else abs(pk) for k, pk in enumerate(signature)]
+            pm = signature[-1]
+            numer = xi0(qs)
+            if pm:
+                numer += (1 if pm > 0 else -1) * xi1(qs)
+            denom = xi0(list(reversed(range(rank))))
+        partition = tuple(map(abs, self.representation.index)) + tuple([0] * self.representation.manifold.rank)
+        monomials_tuples = itertools.chain.from_iterable(
+            more_itertools.distinct_permutations(p) for p in partition_dominance_or_subpartition_cone(partition)
+        )
+        monomials_tuples = filter(lambda p: all(p[i] == 0 or p[i + rank] == 0 for i in range(rank)), monomials_tuples)
+        monomials_tuples = list(monomials_tuples)
+        monomials = [sympy.polys.monomials.Monomial(m, chi_variables).as_expr()
+                     for m in monomials_tuples]
+        chi_coeffs = list(more_itertools.always_iterable(
+            sympy.symbols(' '.join('c{}'.format(i) for i in range(1, len(monomials) + 1)))))
+        exponents = [n % 2 + 1] * len(monomials)  # the correction s.t. chi is the same polynomial for both oddities of n
+        chi_poly = sympy.Poly(sum(c * m**d for c, m, d in zip(chi_coeffs, monomials, exponents)), chi_variables)
+        pr = chi_poly * denom - numer
+        if n % 2:
+            pr = sympy.Poly(pr.subs((g*gc, 1) for g, gc in zip(gammas_sqrt, gammas_conj_sqrt)), chi_variables)
+        else:
+            pr = sympy.Poly(pr.subs((g*gc, 1) for g, gc in zip(gammas, gammas_conj)), chi_variables)
+        sol = list(sympy.linsolve(pr.coeffs(), chi_coeffs)).pop()
+        if n % 2:
+            chi_variables = gammas + gammas_conj
+            chi_poly = sympy.Poly(chi_poly.subs([gr ** 2, g] for gr, g in zip(gammas_sqrt + gammas_conj_sqrt, chi_variables)), chi_variables)
+        p = sympy.Poly(chi_poly.subs((c, c_val) for c, c_val in zip(chi_coeffs, sol)), chi_variables)
+        coeffs = list(map(int, p.coeffs()))
+        monoms = [list(map(int, monom)) for monom in p.monoms()]
+        return coeffs, monoms
+
+    def _inverse(self, X: B.Numeric) -> B.Numeric:
+        return B.inverse(X)
+
+    def num_levels(self) -> int:
+            """Number of levels, L"""
+            return self._num_levels
+
+    def num_eigenfunctions_per_level(self) -> int:
+        """Number of eigenfunctions per level
+        Is it right?"""
+        return self._dimensions
+
+
+class SOCharacter(LieGroupCharacter):
+    def __init__(self, n, signature):
+        self.signature = signature
+        self.n = n
+        self.coeffs, self.monoms = self._load()
+
+    def _load(self):
+        group_name = '{SO}({})'.format(self.n)
+        file_path = Path(__file__).with_name('precomputed_characters.json')
```

Comment:
The __file__ magic seems to be not recommended. refer to this SO (pun intended): https://stackoverflow.com/a/20885799

Maybe there are newer recommendations, please search around